src/include/mempool.h

```diff
@@ -0,0 +1,521 @@
+tracked independently of the pool total.  This additional accounting
+is optional and only incurs an overhead if the debugging is enabled at
+runtime.  This allows developers to see what types are consuming the
+pool resources.
+
+
+Declaring
+---------
+
+Using memory pools is very easy.
+
+To create a new memory pool, simply add a new name into the list of
+memory pools that's defined in "DEFINE_MEMORY_POOLS_HELPER".  That's
+it.  :)
+
+For each memory pool that's created a C++ namespace is also
+automatically created (name is same as in DEFINE_MEMORY_POOLS_HELPER).
+That namespace contains a set of common STL containers that are predefined
+with the appropriate allocators.
+
+Thus for mempool "unittest_1" we have automatically available to us:
+
+   unittest_1::map
+   unittest_1::multimap
+   unittest_1::set
+   unittest_1::multiset
+   unittest_1::list
+   unittest_1::vector
+   unittest_1::unordered_map
+
+
+Putting objects in a mempool
+----------------------------
+
+In order to use a memory pool with a particular type, a few additional
+declarations are needed.
+
+For a class:
+
+  struct Foo {
+    MEMBER_OF_MEMPOOL()
+    ...
+  };
+
+Then, in an appropriate .cc file,
+
+  MEMPOOL_DEFINE_OBJECT_FACTORY(Foo, foo, unittest_1);
+
+The second argument can generally be identical to the first, except
+when the type contains a nested scope.  For example, for
+BlueStore::Onode, we need to do
+
+  MEMPOOL_DEFINE_OBJECT_FACTORY(BlueStore::Onode, bluestore_onode,
+                                bluestore_meta);
+
+(This is just because we need to name some static varables and we
+can't use :: in a variable name.)
+
+In order to use the STL containers, a few additional declarations
+are needed.  For example,
+
+  unittest_1::map<int> myvec;
+
+requires
+
+  MEMPOOL_DEFINE_FACTORY(int, int, unittest_1);
+  MEMPOOL_DEFINE_MAP_FACTORY(int, int, unittest_1);
+
+There are similar macros for SET, LIST, and UNORDERED_MAP.  The MAP
+macro serves both std::map and std::multimap, and std::vector doesn't
+need a second container-specific declaration (because it only
+allocates T and T* arrays; there is no internal container-specific
+wrapper type).
+
+unordered_map is trickier.  First, it has to allocate the hash
+buckets, which requires an extra mempool-wide definition that is
+shared by all different types.  Second, sometimes the hash value is
+cached in the hash node and sometimes it is not.  The glibc STL makes
+its own decision if you don't explicitly define traits, so you either
+need to match your definition with its inference, or explicitly define
+traits, or simply define the allocator for with the cached and
+uncached case and leave one of them unused (but polluting your
+debug dumps).  For example,
+
+  unittest_2::unordered_map<int, std::string> one;       // not cached
+  unittest_2::unordered_map<std::string, int> one;       // cached
+
+needs
+
+ MEMPOOL_DEFINE_UNORDERED_MAP_BASE_FACTORY(unittest_2);  // used by both
```

Comment:
MEMPOOL_DEFINE_UNORDERED_MAP_BASE_FACTORY. Why not ALWAYS declare one of these in the DEFINE_MEMPOOLS_HELPER macro so that people don't have to worry about it? If it's unused, who cares?